File: artifacts/api-server/routers/vendas.py
"""Rotas de Vendas."""
import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from datetime import datetime, date
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from database import get_db
from models import Venda, ItemVenda, Produto, SessaoCaixa, Fiado, Cliente
from schemas import RegistrarVendaInput, VendaOut, ItemVendaOut, ResumoVendas

router = APIRouter()


def fmt_venda(v: Venda, cliente_nome: Optional[str] = None) -> dict:
    return {
        "id": v.id,
        "sessao_id": v.sessao_id,
        "categoria": v.categoria,
        "total": v.total,
        "desconto": v.desconto,
        "pagamento": v.pagamento,
        "cliente_id": v.cliente_id,
        "cliente_nome": cliente_nome or (v.cliente.nome if v.cliente else None),
        "observacao": v.observacao,
        "criado_em": v.criado_em.isoformat() if v.criado_em else None,
    }


@router.get("/resumo/hoje", response_model=ResumoVendas)
def resumo_hoje(db: Session = Depends(get_db)):
    inicio = datetime.combine(date.today(), datetime.min.time())
    fim = datetime.combine(date.today(), datetime.max.time())
    vendas = db.query(Venda).filter(Venda.criado_em >= inicio, Venda.criado_em <= fim).all()

    resumo = dict(total=0, total_dinheiro=0, total_pix=0, total_cartao=0,
                  total_fiado=0, num_vendas=len(vendas), mercado=0, cozinha=0)
    for v in vendas:
        resumo["total"] += v.total
        resumo[f"total_{v.pagamento}"] = resumo.get(f"total_{v.pagamento}", 0) + v.total
        resumo[v.categoria] = resumo.get(v.categoria, 0) + v.total
    return resumo


@router.get("/{id}/itens", response_model=List[ItemVendaOut])
def itens_da_venda(id: int, db: Session = Depends(get_db)):
    itens = db.query(ItemVenda).filter(ItemVenda.venda_id == id).all()
    return [{"id": i.id, "venda_id": i.venda_id, "produto_id": i.produto_id,
             "nome_snap": i.nome_snap, "quantidade": i.quantidade,
             "preco_unit": i.preco_unit, "subtotal": i.subtotal} for i in itens]


@router.get("/", response_model=List[VendaOut])
def listar_vendas(data: Optional[str] = None, data_inicio: Optional[str] = None,
                  data_fim: Optional[str] = None, categoria: Optional[str] = None,
                  limit: int = 50, db: Session = Depends(get_db)):
    query = db.query(Venda)
    if data:
        d = date.fromisoformat(data)
        query = query.filter(
            Venda.criado_em >= datetime.combine(d, datetime.min.time()),
            Venda.criado_em <= datetime.combine(d, datetime.max.time()),
        )
    if data_inicio:
        d = date.fromisoformat(data_inicio)
        query = query.filter(Venda.criado_em >= datetime.combine(d, datetime.min.time()))
    if data_fim:
        d = date.fromisoformat(data_fim)
        query = query.filter(Venda.criado_em <= datetime.combine(d, datetime.max.time()))
    if categoria:
        query = query.filter(Venda.categoria == categoria)
    vendas = query.order_by(Venda.criado_em.desc()).limit(limit).all()
    return [fmt_venda(v) for v in vendas]


@router.post("/", response_model=VendaOut, status_code=201)
def registrar_venda(data: RegistrarVendaInput, db: Session = Depends(get_db)):
    sessao = db.query(SessaoCaixa).filter(SessaoCaixa.status == "aberto").first()
    sessao_id = sessao.id if sessao else None

    subtotal = sum(i.quantidade * i.preco_unit for i in data.itens)
    total = max(0, subtotal - (data.desconto or 0))

    venda = Venda(
        sessao_id=sessao_id,
        categoria=data.categoria,
        total=total,
        desconto=data.desconto or 0,
        pagamento=data.pagamento,
        cliente_id=data.cliente_id,
        observacao=data.observacao,
    )
    db.add(venda)
    db.flush()

    for item in data.itens:
        produto = db.query(Produto).filter(Produto.id == item.produto_id).first()
        if not produto:
            continue
        nome_snap = f"{produto.nome} - {produto.marca}" if produto.marca else produto.nome
        iv = ItemVenda(
            venda_id=venda.id,
            produto_id=item.produto_id,
            nome_snap=nome_snap,
            quantidade=item.quantidade,
            preco_unit=item.preco_unit,
            subtotal=item.quantidade * item.preco_unit,
        )
        db.add(iv)
        produto.estoque = produto.estoque - item.quantidade

    if data.pagamento == "fiado" and data.cliente_id:
        fiado = Fiado(cliente_id=data.cliente_id, venda_id=venda.id, valor=total)
        db.add(fiado)

    if sessao:
        if data.pagamento == "dinheiro":
            sessao.total_dinheiro += total
        elif data.pagamento == "pix":
            sessao.total_pix += total
        elif data.pagamento == "cartao":
            sessao.total_cartao += total
        elif data.pagamento == "fiado":
            sessao.total_fiado += total

    db.commit()
    db.refresh(venda)

    cliente_nome = None
    if venda.cliente_id:
        c = db.query(Cliente).filter(Cliente.id == venda.cliente_id).first()
        cliente_nome = c.nome if c else None

    return fmt_venda(venda, cliente_nome)

File: artifacts/api-server/src/routes/vendas.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import {
  vendasTable,
  itensVendaTable,
  produtosTable,
  sessoesCaixaTable,
  fiadosTable,
  clientesTable,
  nfceLogsTable,
} from "@workspace/db/schema";
import { eq, and, gte, lte, sql, desc, inArray } from "drizzle-orm";
import { RegistrarVendaBody } from "@workspace/api-zod";
import { emitirNfce } from "../services/sefaz.service";
import { imprimirDanfeSimplificado } from "../services/danfe.service";

const router: IRouter = Router();

router.get("/", async (req, res) => {
  const { data, data_inicio, data_fim, categoria, limit } = req.query as {
    data?: string;
    data_inicio?: string;
    data_fim?: string;
    categoria?: string;
    limit?: string;
  };

  const lim = Math.min(1000, Math.max(1, parseInt(limit ?? "500") || 500));
  const conditions = [];
  if (data) {
    const start = new Date(data);
    start.setHours(0, 0, 0, 0);
    const end = new Date(data);
    end.setHours(23, 59, 59, 999);
    conditions.push(gte(vendasTable.criado_em, start));
    conditions.push(lte(vendasTable.criado_em, end));
  }
  if (data_inicio) {
    const start = new Date(data_inicio);
    start.setHours(0, 0, 0, 0);
    conditions.push(gte(vendasTable.criado_em, start));
  }
  if (data_fim) {
    const end = new Date(data_fim);
    end.setHours(23, 59, 59, 999);
    conditions.push(lte(vendasTable.criado_em, end));
  }
  if (categoria) conditions.push(eq(vendasTable.categoria, categoria));

  const vendas = await db
    .select({ venda: vendasTable, cliente_nome: clientesTable.nome })
    .from(vendasTable)
    .leftJoin(clientesTable, eq(vendasTable.cliente_id, clientesTable.id))
    .where(conditions.length ? and(...conditions) : undefined)
    .orderBy(desc(vendasTable.criado_em))
    .limit(lim);

  res.json(
    vendas.map(({ venda, cliente_nome }) => ({
      ...venda,
      cliente_nome,
      criado_em: venda.criado_em.toISOString(),
    })),
  );
});
router.get("/resumo/hoje", async (req, res) => {
  const hoje = new Date();
  hoje.setHours(0, 0, 0, 0);
  const amanha = new Date(hoje);
  amanha.setDate(amanha.getDate() + 1);
  const vendas = await db
    .select()
    .from(vendasTable)
    .where(and(gte(vendasTable.criado_em, hoje), lte(vendasTable.criado_em, amanha)));
  const resumo = {
    total: 0,
    total_dinheiro: 0,
    total_pix: 0,
    total_cartao: 0,
    total_fiado: 0,
    num_vendas: vendas.length,
    mercado: 0,
    cozinha: 0,
    feira: 0,
  };
  for (const v of vendas) {
    resumo.total += v.total;
    if (v.pagamento === "dinheiro") resumo.total_dinheiro += v.total;
    if (v.pagamento === "pix") resumo.total_pix += v.total;
    if (v.pagamento === "cartao") resumo.total_cartao += v.total;
    if (v.pagamento === "fiado") resumo.total_fiado += v.total;
    if (v.categoria === "mercado") resumo.mercado += v.total;
    if (v.categoria === "cozinha") resumo.cozinha += v.total;
    if (v.categoria === "feira") resumo.feira += v.total;
  }
  res.json(resumo);
});
router.get("/:id/itens", async (req, res) => {
  const id = parseInt(req.params.id);
  const itens = await db.select().from(itensVendaTable).where(eq(itensVendaTable.venda_id, id));
  res.json(itens);
});


router.post("/", async (req, res) => {
  const data = RegistrarVendaBody.parse(req.body);
  const rawBody = req.body as Record<string, unknown>;

  const sessao = await db.query.sessoesCaixaTable.findFirst({
    where: eq(sessoesCaixaTable.status, "aberto"),
  });

  const subtotal = data.itens.reduce((acc, item) => acc + item.quantidade * item.preco_unit, 0);
  const total = Math.max(0, subtotal - (data.desconto ?? 0));

  const [venda] = await db.insert(vendasTable)
    .values({
      sessao_id: sessao?.id ?? null,
      categoria: data.categoria,
      total,
      desconto: data.desconto ?? 0,
      pagamento: data.pagamento,
      cliente_id: data.cliente_id ?? null,
      observacao: data.observacao ?? null,
    })
    .returning();

  const insertedItensVenda: (typeof itensVendaTable.$inferSelect)[] = [];
  for (const item of data.itens) {
    const produto = await db.query.produtosTable.findFirst({ where: eq(produtosTable.id, item.produto_id) });
    if (!produto) continue;

    const nome_snap = produto.marca ? `${produto.nome} - ${produto.marca}` : produto.nome;

    const [insertedItem] = await db.insert(itensVendaTable).values({
      venda_id: venda.id,
      produto_id: item.produto_id,
      nome_snap,
      quantidade: item.quantidade,
      unidades: null,
      preco_unit: item.preco_unit,
      subtotal: item.quantidade * item.preco_unit,
    }).returning();
    insertedItensVenda.push(insertedItem);

    const qtdBaixaEstoque = item.quantidade;
    const novoEstoque = produto.estoque - qtdBaixaEstoque;
    await db.update(produtosTable).set({ estoque: novoEstoque }).where(eq(produtosTable.id, item.produto_id));
  }

  if (data.pagamento === "fiado" && data.cliente_id) {
    await db.insert(fiadosTable).values({
      cliente_id: data.cliente_id,
      venda_id: venda.id,
      valor: total,
      pago: false,
    });
  }

  if (sessao) {
    const updates: Record<string, number> = {};
    if (data.pagamento === "dinheiro") updates.total_dinheiro = sql`total_dinheiro + ${total}` as unknown as number;
    if (data.pagamento === "pix") updates.total_pix = sql`total_pix + ${total}` as unknown as number;
    if (data.pagamento === "cartao") updates.total_cartao = sql`total_cartao + ${total}` as unknown as number;
    if (data.pagamento === "fiado") updates.total_fiado = sql`total_fiado + ${total}` as unknown as number;

    if (Object.keys(updates).length > 0) {
      await db.update(sessoesCaixaTable).set(updates).where(eq(sessoesCaixaTable.id, sessao.id));
    }
  }
  
  // Responder imediatamente ao frontend
  res.status(201).json({
    ...venda,
    criado_em: venda.criado_em.toISOString(),
    danfe_impresso: false,
    nfce_status: "pendente",
  });

  // Processar NFC-e em background (não-bloqueante)
  setImmediate(async () => {
    try {
      console.log("[NFC-e background] iniciando para venda", venda.id);
      const produtosIds = data.itens.map((i) => i.produto_id);
      const produtos = await db.query.produtosTable.findMany({ where: inArray(produtosTable.id, produtosIds) });
      const cliente = venda.cliente_id ? await db.query.clientesTable.findFirst({ where: eq(clientesTable.id, venda.cliente_id) }) : null;

      const cpfNota = typeof rawBody.cpf_nota === "string" ? rawBody.cpf_nota.replace(/\D/g, "") : "";
      const vendaParaSefaz = {
        ...venda,
        observacao:
          cpfNota.length === 11
            ? `${venda.observacao ? `${venda.observacao} | ` : ""}CPF_NA_NOTA:${cpfNota}`
            : venda.observacao,
      };

      const emissao = await emitirNfce(vendaParaSefaz, insertedItensVenda, produtos, cliente);

      if (emissao.success && emissao.xmlAutorizado) {
        const printTimeoutMs = Number(process.env.DANFE_PRINT_TIMEOUT_MS || "12000");
        try {
          await Promise.race([
            imprimirDanfeSimplificado(emissao.qrCodeUrl || "", emissao.chaveAcesso || "", emissao.xmlAutorizado),
            new Promise((_, reject) =>
              setTimeout(
                () => reject(new Error(`Timeout da impressao DANFE (${printTimeoutMs}ms)`)),
                printTimeoutMs,
              ),
            ),
          ]);
        } catch (printError) {
          console.error("[NFC-e background] Falha ao imprimir DANFE:", printError);
        }
      }
    } catch (bgErr) {
      console.error("[NFC-e background] erro:", bgErr);
    }
  });
});

router.put("/:id", async (req, res) => {
  const id = Number(req.params.id);
  if (!Number.isFinite(id)) {
    res.status(400).json({ ok: false, message: "id invalido" });
    return;
  }
  const payload = req.body as Partial<{
    pagamento: "dinheiro" | "pix" | "cartao" | "fiado";
    cliente_id: number | null;
    observacao: string | null;
  }>;
  const [updated] = await db
    .update(vendasTable)
    .set({
      pagamento: payload.pagamento,
      cliente_id: payload.cliente_id,
      observacao: payload.observacao,
    })
    .where(eq(vendasTable.id, id))
    .returning();
  if (!updated) {
    res.status(404).json({ ok: false, message: "Venda nao encontrada" });
    return;
  }
  res.json({ ...updated, criado_em: updated.criado_em.toISOString() });
});

export default router;